riscv/csrs.cc

```diff
@@ -1239,7 +1239,7 @@ bool hgatp_csr_t::unlogged_write(const reg_t val) noexcept {
   proc->get_mmu()->flush_tlb();
 
   reg_t mask;
-  if (proc->get_const_xlen() == 32) {
+  if (proc->get_xlen(PRV_S) == 32) {
```

Comment:
This is also one of my question. The fields for HSXLEN=64 cannot totally map to fields for HSXLEN=32, even though they have the same field names. I think this may be the work of software.

Reply:
Yes, what's supposed to happen here when XLEN changes doesn't seem to be well-specified. But at least we could make some effort not to leave total garbage in there -- move VMID, reset Mode, etc?

I think you're pioneering a lot of architectural work here, and there will be many more questions like this that we don't have good answers for.

Reply:
“Complete garbage” might be what’s required by the CSR width modulation section in chapter 2. I think it requires a bitwise copy of the whole register, even if bits end up in the wrong fields.

Reply:
The current implementation is to use bitwise copy for most CSRs, by leaving them alone, and truncating their length to their related XLEN when read and write. 
However, the CSR value may be unchanged if their XLEN is back-to-back changed(64->32->64). This seems a little different from the description from the CSR width modulation section in chapter 2：
_`“If the new width is wider than the previous width, the temporary register is zero-extended to the wider width”`_

Reply:
> “Complete garbage” might be what’s required by the CSR width modulation section in chapter 2. I think it requires a bitwise copy of the whole register, even if bits end up in the wrong fields.

The opening paragraph there says: "fields and bits of the new-width CSR are, unless specified otherwise, determined..."

I think the two different figures for `hgatp` for HSXLEN of 32/64 count as "specified otherwise." VMID and MODE move and it wouldn't be correct to jam the upper bits of PPN into the VMID field when switching from 64 to 32.

This kind of per-field relocation might not be well explained in the spec, but is necessary in order for changing `misa.MXL` back and forth between 2 (xlen=64) and 1 (xlen=32) to work.

But it's because of architectural questions like this that I am very leery of this being the right time & place to pioneer this feature.